app/upgrades/v20/upgrades.go

```diff
@@ -134,133 +136,512 @@ func InitializeLastProviderConsensusValidatorSet(
-func MigrateProposal(ctx sdk.Context, govKeeper govkeeper.Keeper, proposal govtypes.Proposal) error {
-	for idx, msg := range proposal.GetMessages() {
-		sdkLegacyMsg, isLegacyProposal := msg.GetCachedValue().(*govtypes.MsgExecLegacyContent)
-		if !isLegacyProposal {
-			continue
+// MigrateProposal migrates an ICS proposal
+func MigrateProposal(
+	ctx sdk.Context,
+	msgServer providertypes.MsgServer,
+	providerKeeper providerkeeper.Keeper,
+	govKeeper govkeeper.Keeper,
+	proposal govtypes.Proposal,
+) error {
+	// ignore proposals that were rejected or failed
+	if proposal.Status != govtypes.StatusDepositPeriod &&
+		proposal.Status != govtypes.StatusVotingPeriod &&
+		proposal.Status != govtypes.StatusPassed {
+		return nil
+	}
+
+	// ignore proposals with more than one message as we are interested only
+	// in legacy proposals (which have only one message)
+	messages := proposal.GetMessages()
+	if len(messages) != 1 {
+		return nil
+	}
+	msg := messages[0]
+
+	// ignore non-legacy proposals
+	sdkLegacyMsg, isLegacyProposal := msg.GetCachedValue().(*govtypes.MsgExecLegacyContent)
+	if !isLegacyProposal {
+		return nil
+	}
+	content, err := govtypes.LegacyContentFromMessage(sdkLegacyMsg)
+	if err != nil {
+		return err
+	}
+
+	switch msg := content.(type) {
+	case *providertypes.ConsumerAdditionProposal:
+		return MigrateConsumerAdditionProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+
+	case *providertypes.ConsumerRemovalProposal:
+		return MigrateConsumerRemovalProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+
+	case *providertypes.ConsumerModificationProposal:
+		return MigrateConsumerModificationProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+
+	case *providertypes.ChangeRewardDenomsProposal:
+		return MigrateChangeRewardDenomsProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+	}
+
+	return nil
+}
+
+// MigrateConsumerAdditionProposal migrates a ConsumerAdditionProposal
+func MigrateConsumerAdditionProposal(
+	ctx sdk.Context,
+	msgServer providertypes.MsgServer,
+	providerKeeper providerkeeper.Keeper,
+	govKeeper govkeeper.Keeper,
+	proposal govtypes.Proposal,
+	msg *providertypes.ConsumerAdditionProposal,
+) error {
+	if proposal.Status == govtypes.StatusPassed {
+		// ConsumerAdditionProposal that passed -- it was added to the
+		// list of pending consumer addition proposals, which was deleted during
+		// the migration of the provider module
+		for _, consumerId := range providerKeeper.GetAllActiveConsumerIds(ctx) {
+			chainId, err := providerKeeper.GetConsumerChainId(ctx, consumerId)
+			if err != nil {
+				return err // this means is something wrong with the provider state
+			}
+			if chainId == msg.ChainId {
+				// this proposal was already handled in a previous block
```

Comment:
Not sure I get why this is needed? Is the idea that `MigrateProposal` might go to older proposals as well (e.g., when we introduced Stride and Neutron) and skip re-creating those here?

Reply:
Yes. The iteration goes through all proposals, even old ones. 

We want to differentiate between proposals that passed and were not handled yet and those that were already handled. According to the exiting logic (pre v20), when a ConsumerAdditionProposal passes, it is added to a list of pending props. Then, at spawn time, this proposal is handled (i.e., a consumer client and a consumer genesis is created). If the proposal was handled, then the result chain is already migrated during the provider module migration. This means that there should be a mappping from consumerId to chainId.